tracer/src/Datadog.Trace/ClrProfiler/AutoInstrumentation/Grpc/GrpcLegacy/Client/GrpcLegacyClientCommon.cs

```diff
@@ -73,10 +74,15 @@ internal static class GrpcLegacyClientCommon
 
                 span.SetHeaderTags(requestMetadataWrapper, tracer.Settings.GrpcTags, GrpcCommon.RequestMetadataTagPrefix);
                 scope = tracer.ActivateSpan(span);
+                var traceContext = span.Context.TraceContext;
 
-                if (setSamplingPriority && existingSpanContext?.SamplingPriority is not null)
+                if (setSamplingPriority && existingSpanContext?.SamplingPriority is not null && traceContext != null)
                 {
-                    span.SetTraceSamplingPriority(existingSpanContext.SamplingPriority.Value);
+                    traceContext.SetSamplingPriority(existingSpanContext.SamplingPriority);
+
+                    // copy propagated tags
+                    var traceTags = TagPropagation.ParseHeader(existingSpanContext.PropagatedTags);
+                    traceContext.Tags.SetTags(traceTags);
```

Comment:
Could we do all that at span creation (inside a new Tracer.StartSpan internal overload where we would pass the extracted context) to avoid doing this in each integration?

Reply:
> Related question I assume, why don't we do this for all integrations where we extract context from headers?

This is a tricky situation, where we have a "split" `Start` and `End` async methods, where we can't guarantee that the `End` method will be called (and if it's not, it could have Very Bad consequences as we leak the span. In the case of GRPC, IIRC, it's because we need to inject the propagation headers in the "start" method, but we can't create the span until the "end" method.

The "Start" instrumentation records everything that's needed to recreate the span later, the End instrumentation then calls this `CreateClientSpan` method, reuses the values it injected into the outgoing headers, and creates the spans. 

I think currently there's a bug, in that if we are meant to be propagating the sampling decision, that needs to be done elsewhere. The `Span` created in this method is _not_ used for propagation.

Reply:
I commented on the HTTP one, but I'll comment here too: Actually, since this tag is one of the PropagatedTags, I think it is already handled when we inject the inactive span into the headers.

On line 148 of this file we inject the current span context into the headers, so the propagated tags get added automatically

Reply:
Oh, nice. Thank you! I removed the code that copies that individual trace tags over.